jax/interpreters/pxla.py

```diff
@@ -579,9 +648,15 @@ def dynamic_fun(dummy, *args):
   compiled = built.Compile(compile_options=compile_options,
                            backend=xb.get_backend(backend))
 
-  handle_args = partial(shard_args, compiled.local_devices(),
-                        assign_shards_to_replicas(num_local_replicas, axis_size),
-                        axis_size)
+  sharding_specs = [_pmap_sharding_spec(num_local_replicas, axis_size,
+                                        aval.shape)
+                    if aval is not core.abstract_unit else None
+                    for aval in sharded_avals]
+  indices = [spec_to_indices(aval.shape, spec)
+             if spec is not None else None
+             for aval, spec in zip(avals, sharding_specs)]
+  handle_args = partial(shard_args, compiled.local_devices(), indices)
```

Comment:
This is where we're equating the shard spec's shard order to the (linearized) device assignment, right? Meaning that the order of logical array axes in JAX would determine which axes are major or minor in the physical topology?
(Also, for clarification,: does `compiled.local_devices()` return the local devices in the linearized order of the XLA logical (replica, partition) they're each mapped to?)

Reply:
Correct. This is also important for things like calculating replica groups. Ideally we could make this super important and subtle invariant more explicit somehow. I don't have any ideas off the top of my head but I bet we can come up with something. This might also help with the headaches I get whenever I have to think about the 3-way data<>replica<>physical mapping...

`compiled.local_devices()` does return them in replica x partition linearized order, although it's not documented anywhere! :open_mouth: It happens here: https://github.com/tensorflow/tensorflow/blob/master/tensorflow/compiler/xla/python/local_client.cc#L994